util/include/Serializable.h

```diff
@@ -37,72 +40,199 @@ typedef std::unique_ptr<char[], std::default_delete<char[]>> UniquePtrToChar;
+    std::copy(vec.begin(), vec.end(), std::inserter(set, set.begin()));
+  }
+  /** *****************************************************************************************************************
+   *  Serializable
+   *  - name
+   *  - version
+   *  - data members (recursively)
+   */
+  template<typename T, typename std::enable_if<std::is_convertible<T*, Serializable*>::value>::type* = nullptr>
+  static void serialize_impl(std::ostream& outStream, const T& t, int) {
+    const Serializable& s = static_cast<const Serializable&>(t);
+    serialize(outStream, s.getName());
+    serialize(outStream, s.getVersion());
+    s.serializeDataMembers(outStream);
+  }
+  template<typename T, typename std::enable_if<std::is_convertible<T, Serializable*>::value>::type* = nullptr>
+  static void serialize_impl(std::ostream& outStream, const T t, int) {
+    const Serializable& s = static_cast<const Serializable&>(*t);
+    serialize_impl(outStream, s, int{});
+  }
+  template<typename T, typename std::enable_if<std::is_convertible<T, Serializable*>::value>::type* = nullptr>
+  static void deserialize_impl(std::istream& inStream, T& t, int) {
+    std::string className;
+    deserialize(inStream, className);
+    LOG_TRACE(logger(), className.c_str());
+    auto it = registry().find(className);
+    if (it == registry().end())
+      throw std::runtime_error("Deserialization failed: unknown class name: " + className);
+    Serializable* s = it->second();
+    std::string version;
+    deserialize(inStream, version);
+    if (version != s->getVersion())
+      throw std::runtime_error("Deserialization failed: wrong version: " +  version +
+                               std::string(", expected version: ") + s->getVersion());
+    s->deserializeDataMembers(inStream);
+
+    t = dynamic_cast<T>(s);// shouldn't throw because T is convertible to Serializable*
+  }
+  template<typename T, typename std::enable_if<std::is_convertible<T*, Serializable*>::value>::type* = nullptr>
+  static void deserialize_impl(std::istream& inStream, T& t, int) {
+    T* s = nullptr;
+    deserialize_impl(inStream, s, int{});
+    t = T(*s);
+  }
+  /** *****************************************************************************************************************
+   *  std::string
+   *  - string length
+   *  - char array
+   */
+  template<typename T, typename std::enable_if<std::is_same<T, std::string>::value>::type* = nullptr>
+  static void serialize_impl(std::ostream& outStream, const T& str, int) {
+    LOG_TRACE(logger(), str);
+    const std::size_t sz = str.size();
+    serialize_impl(outStream, sz, int{});
+    std::string::size_type size = str.size();
+    serialize(outStream, str.data(), size);
+  }
+  template<typename T, typename std::enable_if<std::is_same<T, std::string>::value>::type* = nullptr>
+  static void deserialize_impl(std::istream& inStream, T& t, int) {
+    std::size_t sz;
+    deserialize_impl(inStream, sz, int{});
+    char* str =  new char[sz];
```

Comment:
Is it possible to avoid the copy and read directly into the string?

Reply:
I don't know a good way of doing it.

Reply:
Me neither :)